spring-cloud-task-integration-tests/src/test/java/org/springframework/cloud/task/listener/BatchExecutionEventTests.java

```diff
@@ -0,0 +1,252 @@
+ *  * You may obtain a copy of the License at
+ *  *
+ *  *      http://www.apache.org/licenses/LICENSE-2.0
+ *  *
+ *  * Unless required by applicable law or agreed to in writing, software
+ *  * distributed under the License is distributed on an "AS IS" BASIS,
+ *  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  * See the License for the specific language governing permissions and
+ *  * limitations under the License.
+ *
+ */
+package org.springframework.cloud.task.listener;
+
+import configuration.JobConfiguration;
+import org.junit.After;
+import org.junit.ClassRule;
+import org.junit.Test;
+import org.springframework.batch.core.JobExecution;
+import org.springframework.batch.core.StepExecution;
+import org.springframework.batch.core.scope.context.ChunkContext;
+import org.springframework.boot.autoconfigure.EnableAutoConfiguration;
+import org.springframework.boot.autoconfigure.PropertyPlaceholderAutoConfiguration;
+import org.springframework.boot.autoconfigure.batch.BatchAutoConfiguration;
+import org.springframework.boot.builder.SpringApplicationBuilder;
+import org.springframework.cloud.stream.annotation.EnableBinding;
+import org.springframework.cloud.stream.annotation.StreamListener;
+import org.springframework.cloud.stream.binder.redis.config.RedisServiceAutoConfiguration;
+import org.springframework.cloud.stream.messaging.Sink;
+import org.springframework.cloud.stream.test.junit.redis.RedisTestSupport;
+import org.springframework.cloud.task.batch.configuration.TaskBatchAutoConfiguration;
+import org.springframework.cloud.task.batch.listener.BatchEventAutoConfiguration;
+import org.springframework.context.ConfigurableApplicationContext;
+import org.springframework.context.annotation.PropertySource;
+import org.springframework.messaging.Message;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertTrue;
+
+/**
+ * @author Glenn Renfro
+ */
+public class BatchExecutionEventTests {
+	@ClassRule
+	public static RedisTestSupport redisTestSupport = new RedisTestSupport();
+
+	// Count for two job execution events per job
+	static CountDownLatch jobExecutionLatch = new CountDownLatch(2);
+
+	// Count for four step execution events per job
+	static CountDownLatch stepExecutionLatch = new CountDownLatch(4);
+
+	// Count for twelve item process events per job
+	static CountDownLatch itemProcessLatch = new CountDownLatch(6);
+
+	// Count for eight chunk events per job
+	static CountDownLatch chunkEventsLatch = new CountDownLatch(8);
+
+	// Count for six read events per job
+	static CountDownLatch itemReadEventsLatch = new CountDownLatch(6);
+
+	// Count for six write events per job
+	static CountDownLatch itemWriteEventsLatch = new CountDownLatch(2);
```

Comment:
This countdown latch approach is ok, but it runs the risk of not testing for multiple messages.  It's basically saying "I received at least 2 messages" instead of saying "I received exactly 2 messages"...